### PR TITLE
feature: auth client

### PR DESCRIPTION
**Pull request by LingKa28**
Base on #6 

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Add new feature.

* what changes does this pull request make?
  * add auth client.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
No.

### CODE REVIEW

File: client/auth.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        Grant role for an user.
+        """
+        req = AuthUserGrantRoleRequest(user=user, role=role)
+        request_with_token = RequestWithToken(token=self.token, auth_user_grant_role_request=req)
+        er = await self.handle_req(req=request_with_token, use_fast_path=False)
+        return er.auth_user_grant_role_response
+
+    async def user_revoke_role(self, name: str, role: str) -> AuthUserRevokeRoleResponse:
+        """
+        Revoke role for an user.
```
[the remark this answers is not in the document]
Comment (rogercloud) [reply client/auth.py:181]: `a user` instead of `an user`.

File: client/auth.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+        return er.auth_user_revoke_role_response
+
+    async def role_add(self, name: str) -> AuthRoleAddResponse:
+        """
+        Adds role.
+        """
+        req = AuthRoleAddRequest(name=name)
+        if req.name == "":
+            msg = "role name is empty"
+            raise Exception(msg)
```
Comment (rogercloud) [inline client/auth.py:195]: Please check before constructing the request.

Comment (rogercloud) [reply client/auth.py:195]: Please double-check the problem in the whole code base.

File: client/auth.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        request_with_token = RequestWithToken(token=self.token, auth_role_list_request=AuthRoleListRequest())
+        er = await self.handle_req(req=request_with_token, use_fast_path=True)
+        return er.auth_role_list_response
+
+    async def role_delete(self, role: str) -> AuthRoleDeleteResponse:
+        """
+        Deletes role.
+        """
+        req = AuthRoleDeleteRequest(role=role)
+        request_with_token = RequestWithToken(token=self.token, auth_role_delete_request=req)
```
[the remark this answers is not in the document]
Comment (rogercloud) [reply client/auth.py:222]: There're other unnecessary local variables. Please double check.

File: tests/auth_test.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
+"""Tests for the auth client"""
```
Comment (rogercloud) [inline tests/auth_test.py:1]: We should also check the failure case. Please add the failure cases.

File: client/auth.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        er = await self.handle_req(req=request_with_token, use_fast_path=False)
+        return er.auth_user_delete_response
+
+    async def user_change_password(self, name: str, password: str) -> AuthUserChangePasswordResponse:
+        """
+        Change password for an user.
+        """
+        req = AuthUserChangePasswordRequest(name=name, password=password)
+        if req.password == "":
+            msg = "role name is empty"
```
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:160]: Done.

File: client/auth.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:181]: Done.

File: client/auth.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+        Revoke role permission.
+        """
+        req = AuthRoleRevokePermissionRequest(role=role, key=key, range_end=range_end)
+        request_with_token = RequestWithToken(token=self.token, auth_role_revoke_permission_request=req)
+        er = await self.handle_req(req=request_with_token, use_fast_path=False)
+        return er.auth_role_revoke_permission_response
+
+    async def handle_req(self, req: RequestWithToken, use_fast_path: bool) -> CommandResponse:
+        """
+        Send request using fast path.
```
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:255]: Done.

File: client/auth.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+    async def user_add(self, name: str, password: str = "", no_password: bool = False) -> AuthUserAddResponse:
+        """
+        Add an user.
+        """
+        req = AuthUserAddRequest(name=name, password=password, options=UserAddOptions(no_password=no_password))
+        if req.name == "":
+            msg = "user name is empty"
+            raise Exception(msg)
+        need_password = req.options.no_password
```
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:116]: There is indeed a problem with the logic here. I have corrected.

File: client/auth.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+        req.hashedPassword = hashed_password
+        req.password = ""
+
+        request_with_token = RequestWithToken(token=self.token, auth_user_change_password_request=req)
+        er = await self.handle_req(req=request_with_token, use_fast_path=False)
+        return er.auth_user_change_password_response
+
+    async def user_grant_role(self, user: str, role: str) -> AuthUserGrantRoleResponse:
+        """
+        Grant role for an user.
```
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:172]: Done.

File: client/auth.py (hunk 2 above)
Comment (LingKa28) [reply client/auth.py:195]: I've modified a similar problem that exists whole code.

File: client/auth.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:222]: I've modified a similar problem that exists whole the code

File: client/client.py [hunk 9]
```diff
@@ -1,30 +1,37 @@
...
         self.kv_client = kv
+        self.auth_client = auth
 
     @classmethod
     async def connect(cls, addrs: list[str]) -> Client:
         """
         New `Client`
         """
         protocol_client = await ProtocolClient.build_from_addrs(addrs)
+        channel = grpc.aio.insecure_channel(addrs[0])
```
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/client.py:32]: Yes. It'll be done before our next release.

[the remark this answers is not in the document]
Comment (LingKa28) [reply client/client.py:32]: It seems like we are not support it on the server side at the moment. Ww should wait until the Xline supports `tls`

File: client/auth.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+        if req.name == "":
+            msg = "user name is empty"
+            raise Exception(msg)
+        need_password = req.options.no_password
+        if need_password and req.password == "":
+            msg = "Password is required but not provided"
+            raise Exception(msg)
+        hashed_password = self.hash_password(req.password)
+        req.hashedPassword = hashed_password
+        req.password = ""
```
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:122]: It's seems to be for security reasons. Rust clinet does the same thing.

File: client/auth.py [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+        """
+        Change password for an user.
+        """
+        req = AuthUserChangePasswordRequest(name=name, password=password)
+        if req.password == "":
+            msg = "role name is empty"
+            raise Exception(msg)
+        hashed_password = self.hash_password(req.password)
+        req.hashedPassword = hashed_password
+        req.password = ""
```
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:164]: Same as above

File: client/auth.py [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+    def generate_propose_id(self) -> str:
+        """Generate propose id with the given prefix."""
+        propose_id = f"{self.name}-{uuid.uuid4()}"
+        return propose_id
+
+    @staticmethod
+    def hash_password(password: str) -> str:
+        """Generate hash of the password."""
+        hashed_password = pbkdf2_sha256.using(rounds=200000, salt_size=16).hash(password)
```
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/auth.py:278]: Done.

File: client/auth.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+        er = await self.handle_req(req=request_with_token, use_fast_path=False)
+        return er.auth_role_delete_response
+
+    async def role_grant_permission(
+        self, name: str, perm_type: PermissionType, key: bytes, range_end: bytes | None = None
+    ) -> AuthRoleGrantPermissionResponse:
+        """
+        Grants role permission.
+        """
+        if perm_type is None:
```
Comment (bsbds) [inline client/auth.py:243]: The permission type cannot be None.

Comment (LingKa28) [reply client/auth.py:243]: You are right. param `perm_type` and `key` isn't option.

**Review by bsbds — APPROVED** [review]

**Pull request merged**